### DISCUSSION TITLE
How to Correctly Handle a Model with Concatenation Layer

### DISCUSSION THREAD

**Discussion by williamliao28** [Xilinx/finn#845] [Q&A] [answered]
### Any suggestions are welcomed. I have tried the following experiments for this issue.

First, I have looked into the source code of the `InferConcatLayer()` function (line 1678 to 1737 in `convert_to_hls_layers.py`). I think this function is used to convert the `Concat` ONNX layer to the HLS layer `StreamingConcat`. I found that it already calls InferDataTypes() on line 1736. So is it necessary to call another `InferDataTypes()` again after calling `InferConcatLayer()`?

Second, I try to call the `InferConcatLayer()` function before `InferBinaryMatrixVectorActivation()` and `InferQuantizedMatrixVectorActivation()`. However, I get the following error message: `AssertionError: Thresholds in MatrixVectorActivation_MatMul_4 can't be expressed with type INT10`. The traceback message is included in the screenshot posted below.

Finally, I also included two example python codes in the links below which may be useful for reproducing the error. The file `simple_test.py` contains the PyTorch module defining the model with Brevitas quantization. The file `simple_test_finn_synthesis.py` is the code I run inside the FINN docker and this one produce the error message described above. The file `simple_test_cifar10_8bit_2epoch.onnx` is a pertained model which can be used directly as the input for `simple_test_finn_synthesis.py`. 

<img width="1529" alt="finn_traceback" src="https://github.com/Xilinx/finn/assets/20777211/707e9ae8-ef2c-46ec-a0ba-0c3946340bcb">

[simple_test.py](https://drive.google.com/file/d/1Ovqbr6RofT1PqdMttqBORPgn7DYY-XOS/view?usp=sharing)

[simple_test_finn_synthesis.py](https://drive.google.com/file/d/1x-a2_pLudfeyeExVYQ-FIVQLwhmnqskV/view?usp=sharing)

[simple_test_cifar10_8bit_2epoch.onnx](https://drive.google.com/file/d/1HsKHmVSbe6rUKOo_VbgHnjrSCu265V2j/view?usp=sharing)

**Comment by auphelia** [comment]
Hi @williamliao28 ,

Thanks for providing additional material to reproduce! I think this is not a question that can be quickly answered.
So, could you please create an [issue](https://github.com/Xilinx/finn/issues) from this discussion with the provided template?

**Comment by williamliao28** [comment]
Hi @auphelia,

Thanks for the reply. I have created an issue (#853 ).

**Reply by auphelia** [reply, marked as answer]
The created issue (https://github.com/Xilinx/finn/issues/853) is closed by this PR https://github.com/Xilinx/finn/pull/857
